include/pybind11/detail/common.h

```diff
@@ -824,10 +822,12 @@ struct is_template_base_of_impl {
 /// Check if a template is the base of a type. For example:
 /// `is_template_base_of<Base, T>` is true if `struct T : Base<U> {}` where U can be anything
 template <template <typename...> class Base, typename T>
-#if !defined(_MSC_VER)
+#if defined(PYBIND11_CPP20) || !defined(_MSC_VER)                                                 \
+    || _MSC_VER > 1930 /* Increment as needed, let MS know please. */
+//     https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros
```

Comment:
A little confused, what's the current MSVC number, 1930 or 1931? This isn't something we have to increment every MSVC release, is it? If it doesn't work now, we should assume we don't know when it will start working. Once it does, then we can set this number. Only put real, known limits, not "I hope they fix this" numbers. I"m guessing it is fine in C++20 mode, though - we test on MSVC C++20?

Reply:
Changed as suggested.

Current is 1930.

We have 3 jobs total with C++20:
* Clang 10 x64
* GCC 10 x64
* MSVC 2019 x86

I think I overreacted. I was a pretty unhappy to discover that even MSVC 2019 and 2022 ICE without this workaround. Thinking: heck, if we didn't mask it here for any MSVC, it may have gotten fixed in all those years. But at second thought, my idea to force us to increment the version with every MSVC release — until there is finally a fix — will backfire badly, because it'll force people to update pybind11, or patch locally, to be able to use a new MSVC release.

Another idea is to gate on `_MSC_VER >= 1940`, with the idea to catch ICEs for the next major update of MSVC, but who knows, really, what the correct number will be? — Staying away from it.

Reply:
Yes. Do we know it works on C++20? The old version simply always used the workaround, the new one seems to think MSVC will be happy with this on C++20 mode. Is it?

Reply:
We don't know before or after my change. The old version was guilty (needs workaround) until proven innocent, which has the drawback of covering up flaws even for new compilers. The new is innocent until proven guilty. I'm trying it out now. My first attempt seems to only kind-of work:
```
Run cmake -S . -B . -DPYBIND11_WERROR=ON -DDOWNLOAD_CATCH=ON -DDOWNLOAD_EIGEN=ON -DCMAKE_CXX_STANDARD=11 -DCMAKE_CXX_STANDARD=20
-- Building for: Visual Studio 17 2022
```
If it doesn't get too messy we can fold it in here.

Reply:
eigen is acting up:
```
test_eigen.cpp
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/Matrix.h([46](https://github.com/pybind/pybind11/runs/5179409281?check_suite_focus=true#step:11:46),40): error C2220: the following warning is treated as an error [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/PlainObjectBase.h(98): message : see reference to class template instantiation 'Eigen::internal::traits<Derived>' being compiled [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
          with
          [
              Derived=Eigen::Matrix<double,-1,-1,0,-1,-1>
          ]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/Matrix.h(180): message : see reference to class template instantiation 'Eigen::PlainObjectBase<Eigen::Matrix<double,-1,-1,0,-1,-1>>' being compiled [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\tests\test_eigen.cpp(39): message : see reference to class template instantiation 'Eigen::Matrix<double,-1,-1,0,-1,-1>' being compiled [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/Matrix.h(46,40): warning C[50](https://github.com/pybind/pybind11/runs/5179409281?check_suite_focus=true#step:11:50)54: operator '&': deprecated between enumerations of different types [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/Matrix.h(46,40): message : to simplify migration, consider the temporary use of /Wv:18 flag with the version of the compiler with which you used to build without warnings [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/TriangularMatrix.h(211,29): warning C5054: operator '&': deprecated between enumerations of different types [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/TriangularMatrix.h(211,29): message : to simplify migration, consider the temporary use of /Wv:18 flag with the version of the compiler with which you used to build without warnings [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\tests\test_eigen.cpp(110): message : see reference to class template instantiation 'Eigen::TriangularView<const MatrixType,1>' being compiled [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
          with
          [
              MatrixType=Eigen::Matrix<double,-1,-1,1,-1,-1>
          ]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/TriangularMatrix.h(212,29): warning C5054: operator '&': deprecated between enumerations of different types [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/TriangularMatrix.h(212,29): message : to simplify migration, consider the temporary use of /Wv:18 flag with the version of the compiler with which you used to build without warnings [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/TriangularMatrix.h(213,29): warning C5054: operator '&': deprecated between enumerations of different types [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/TriangularMatrix.h(213,29): message : to simplify migration, consider the temporary use of /Wv:18 flag with the version of the compiler with which you used to build without warnings [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/TriangularMatrix.h(214,29): warning C5054: operator '&': deprecated between enumerations of different types [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\_deps\eigen-src\Eigen\src/Core/TriangularMatrix.h(214,29): message : to simplify migration, consider the temporary use of /Wv:18 flag with the version of the compiler with which you used to build without warnings [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\include\pybind11\cast.h(1014,1): error C27[52](https://github.com/pybind/pybind11/runs/5179409281?check_suite_focus=true#step:11:52): 'pybind11::detail::type_caster<Eigen::Ref<Eigen::Matrix<double,-1,-1,0,-1,-1>,0,Eigen::OuterStride<-1>>,void>': more than one partial specialization matches the template argument list [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
```

Reply:
The old version had a harmless workaround that might not be needed. The new version has a probability of breaking things that would have worked. We try to guess the safest path unless we test it to see it work. If we can test it and it works in C++20 mode, great! Otherwise err on the side of caution.

Reply:
A nice developer way to disable workarounds might be nice, but all at once would be too much. Maybe a standard searchable comment?

Reply:
> Otherwise err on the side of caution.

Idk, I always cared about helping new compilers getting better. It diminishes the benefit of beta testing-compilers if we cover up anyway.